firedrake/mg/ufl_utils.py

```diff
@@ -156,41 +170,52 @@ def coarsen_function_space(V, self, coefficient_mapping=None):
     except ValueError:
         # Not in an add_hooks context
         pass
-
     return V
 
 
+@coarsen.register(firedrake.Cofunction)
 @coarsen.register(firedrake.Function)
 def coarsen_function(expr, self, coefficient_mapping=None):
     if coefficient_mapping is None:
         coefficient_mapping = {}
     new = coefficient_mapping.get(expr)
     if new is None:
-        V = expr.function_space()
-        manager = firedrake.dmhooks.get_transfer_manager(expr.function_space().dm)
         V = self(expr.function_space(), self)
         new = firedrake.Function(V, name="coarse_%s" % expr.name())
-        manager.inject(expr, new)
-    return new
-
-
-@coarsen.register(firedrake.Constant)
-def coarsen_constant(expr, self, coefficient_mapping=None):
-    if coefficient_mapping is None:
-        coefficient_mapping = {}
-    new = coefficient_mapping.get(expr)
-    if new is None:
-        mesh = self(extract_unique_domain(expr), self)
-        new = firedrake.Constant(numpy.zeros(expr.ufl_shape,
-                                             dtype=expr.dat.dtype),
-                                 domain=mesh)
-        # Share data pointer
-        new.dat = expr.dat
+        expr._child = weakref.proxy(new)
     return new
 
 
 @coarsen.register(firedrake.NonlinearVariationalProblem)
 def coarsen_nlvp(problem, self, coefficient_mapping=None):
+    if hasattr(problem, "_coarse"):
+        return problem._coarse
+
+    def inject_on_restrict(fine, restriction, rscale, injection, coarse):
+        from firedrake.bcs import DirichletBC
+        manager = firedrake.dmhooks.get_transfer_manager(fine)
+        finectx = firedrake.dmhooks.get_appctx(fine)
+        coefficients = finectx.F.coefficients() + finectx.J.coefficients()
+        if finectx.Jp is not None:
+            coefficients.extend(finectx.Jp.coefficients())
+        coefficients = tuple(OrderedDict.fromkeys(coefficients))
+        for fine in coefficients:
+            if hasattr(fine, '_child'):
+                coarse = fine._child.__repr__.__self__
```

Comment:
@connorjward @wence- what is the proper way of getting the strong reference to the `weakref.proxy` defined on line 185? The reason I do this is because we later do type checking inside `inject` and `restrict`, and otherwise the type is `weakref.proxy`.

Reply:
I'm not sure that it's possible to do this. weakrefs can be quite fussy. Are you sure that weakref is the right thing to be doing here?

Reply:
The alternative is to have a different mechanism for type checking, one that does not rely on the actual type, but on an interface. Since Cofunction are one-forms, the type is encoded in `hasattr(fine._child, "arguments")`.

Reply:
This seems like a code smell to me. Is there no other way to know whether or not we have a `Function` or `Cofunction` here?

Reply:
Yes, we are now using `ufl.duals.is_dual`.